runtime/near-vm-logic/src/receipt_manager.rs

```diff
@@ -0,0 +1,493 @@
+                        allowance,
+                        receiver_id: receiver_id.into(),
+                        method_names: method_names
+                            .into_iter()
+                            .map(|method_name| {
+                                String::from_utf8(method_name)
+                                    .map_err(|_| HostError::InvalidMethodName)
+                            })
+                            .collect::<std::result::Result<Vec<_>, _>>()?,
+                    }),
+                },
+            }),
+        );
+        Ok(())
+    }
+
+    /// Attach the [`DeleteKeyAction`] action to an existing receipt.
+    ///
+    /// # Arguments
+    ///
+    /// * `receipt_index` - an index of Receipt to append an action
+    /// * `public_key` - a public key for an access key to delete
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `receipt_index` does not refer to a known receipt.
+    pub fn append_action_delete_key(
+        &mut self,
+        receipt_index: u64,
+        public_key: Vec<u8>,
+    ) -> ExtResult<()> {
+        self.append_action(
+            receipt_index,
+            Action::DeleteKey(DeleteKeyAction {
+                public_key: PublicKey::try_from_slice(&public_key)
+                    .map_err(|_| HostError::InvalidPublicKey)?,
+            }),
+        );
+        Ok(())
+    }
+
+    /// Attach the [`DeleteAccountAction`] action to an existing receipt
+    ///
+    /// # Arguments
+    ///
+    /// * `receipt_index` - an index of Receipt to append an action
+    /// * `beneficiary_id` - an account id to which the rest of the funds of the removed account will be transferred
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `receipt_index` does not refer to a known receipt.
+    pub fn append_action_delete_account(
+        &mut self,
+        receipt_index: u64,
+        beneficiary_id: AccountId,
+    ) -> ExtResult<()> {
+        self.append_action(
+            receipt_index,
+            Action::DeleteAccount(DeleteAccountAction { beneficiary_id }),
+        );
+        Ok(())
+    }
+
+    /// Distribute the gas among the scheduled function calls that specify a gas weight.
+    ///
+    /// Distributes the gas passed in by splitting it among weights defined in `gas_weights`.
+    /// This will sum all weights, retrieve the gas per weight, then update each function
+    /// to add the respective amount of gas. Once all gas is distributed, the remainder of
+    /// the gas not assigned due to precision loss is added to the last function with a weight.
+    ///
+    /// # Arguments
+    ///
+    /// * `gas` - amount of unused gas to distribute
+    ///
+    /// # Returns
+    ///
+    /// Function returns a [GasDistribution] that indicates how the gas was distributed.
+    #[cfg(feature = "protocol_feature_function_call_weight")]
+    pub fn distribute_unused_gas(&mut self, gas: u64) -> GasDistribution {
```

Comment:
Let's name the `gas` here something else -- we have a local `gas` variable inside, and, in this case, this is confusing.